fem/mgbilinearform.cpp

```diff
@@ -0,0 +1,128 @@
+// Copyright (c) 2010, Lawrence Livermore National Security, LLC. Produced at
+// the Lawrence Livermore National Laboratory. LLNL-CODE-443211. All Rights
+// reserved. See file COPYRIGHT for details.
+//
+// This file is part of the MFEM library. For more information and source code
+// availability see http://mfem.org.
+//
+// MFEM is free software; you can redistribute it and/or modify it under the
+// terms of the GNU Lesser General Public License (as published by the Free
+// Software Foundation) version 2.1 dated February 1999.
+
+#include "mgbilinearform.hpp"
+#include "transfer.hpp"
+
+namespace mfem
+{
+
+MultigridBilinearForm::MultigridBilinearForm() : MultigridOperator() {}
+
+MultigridBilinearForm::MultigridBilinearForm(SpaceHierarchy& spaceHierarchy,
+                                             BilinearForm& bf,
+                                             Array<int>& ess_bdr,
+                                             int chebyshevOrder)
+   : MultigridOperator()
+{
+   MFEM_VERIFY(bf.GetAssemblyLevel() == AssemblyLevel::PARTIAL,
+               "Assembly level must be PARTIAL");
+
+   MFEM_VERIFY(bf.GetBBFI()->Size() == 0
+               && bf.GetFBFI()->Size() == 0
+               && bf.GetBFBFI()->Size() == 0,
+               "Only domain integrators are currently supported");
+
+   BilinearForm* form = new BilinearForm(&spaceHierarchy.GetFESpaceAtLevel(0));
+   Array<BilinearFormIntegrator*>& dbfi = *bf.GetDBFI();
+   for (int i = 0; i < dbfi.Size(); ++i)
+   {
+      form->AddDomainIntegrator((*bf.GetDBFI())[i]->Copy());
+   }
```

Comment:
Thank you, I did not see the `seExternalIntegrators()` method. I will try this out instead of using my `Copy()` methods.

Reply:
I strongly agree. I don't like the `Copy()` at all.

Reply:
I remember now why I added the `Copy` methods. I actually tried it with `UseExternalIntegrators` before but that does not work for partially assembled integrators. These integrators have their own member variables which store the pre-computed values. The pre-computed values depend on the finite element space and can not be re-used for finite element spaces of different orders (in the multigrid hierarchy). This means, the integrators cannot be re-used with `UseExternalIntegrators` since they have their finite element space dependent state. I agree with you and do not like the `Copy` methods myself, but I do not see another solution for this issue right now.

Reply:
This is a good point. There are ways to avoid the use of `Copy` but they will probably require more changes than just adding `Copy`. There is a similar method for class `Element` called `Duplicate`; another possible name is `Clone`.

Reply:
This is actually confusing. If you copy the "skeleton" of the integrator but not the particular data it may contain, then should you call it "copy", "clone", or "duplicate". All those seem confusing. This is a form of a "shallow copy" or even "transfer" of the integrator, or "recreate". This is likely to be something that would bite back at us in the future.

Maybe it's too late but one way would have been to have "context"/"state" of the integrator stored separately from the integrator and invoking the integrator with the provision of the context. Then, it would be a context switch every time you reuse it, as it is a common way to use a shared resource.

Reply:
You are right @delyank. The name was very confusing. I decided to move the (Par)MultigridBilinearForm classes to the examples and to get rid of the Copy methods. Independently of this pull request, it might be necessary to add these methods again in the future, but this should be done in a separate pull request/discussion.